mountpoint-s3/src/inode.rs

```diff
@@ -131,63 +131,65 @@ impl Superblock {
     /// The kernel may forget a number of references (`n`) in one forget message to our FUSE implementation.
     /// If the lookup count reaches zero, it is safe for the [Superblock] to delete the [Inode].
     pub fn forget(&self, ino: InodeNo, n: u64) {
-        // Put inode removal in a block so we don't hold the lock on the inodes table longer than needed.
-        let (inode, parent) = {
-            let mut inodes = self.inner.inodes.write().unwrap();
-            match inodes.get(&ino) {
-                None => {
-                    debug_assert!(
-                        false,
-                        "forget should not be called on inode already removed from superblock"
-                    );
-                    error!("forget called on inode {ino} already removed from the superblock");
-                    return;
-                }
-                Some(inode) => {
-                    logging::record_name(inode.name());
-                    let new_lookup_count = inode.dec_lookup_count(n);
-                    if new_lookup_count == 0 {
-                        // Safe to remove, kernel no longer has a reference to it.
-                        trace!(ino, "removing inode from superblock");
-                        let inode = inodes.remove(&ino).expect("we just retrieved it");
-
-                        // Should be impossible for this to fail (VFS inodes reference their parent, so
-                        // children need to be freed first), but let's not crash in a `forget` function...
-                        let Some(parent) = inodes.get(&inode.parent()) else {
-                            debug_assert!(false, "children should be forgotten before parents");
-                            return;
-                        };
-                        (inode, parent.clone())
-                    } else {
+        let inodes = self.inner.inodes.read().unwrap();
+        match inodes.get(&ino) {
+            None => {
+                debug_assert!(
+                    false,
+                    "forget should not be called on inode already removed from superblock"
+                );
+                error!("forget called on inode {ino} already removed from the superblock");
+            }
+            Some(inode) => {
+                // We have to clone the inode so it will not hold reference to the lock which we want to drop as soon as possible.
+                let inode = inode.clone();
+                drop(inodes);
```

Comment:
I'd recommend something like this (untested):

```suggestion
        match self.inner.inodes.read().unwrap().get(&ino).cloned() {
            None => {
                debug_assert!(
                    false,
                    "forget should not be called on inode already removed from superblock"
                );
                error!("forget called on inode {ino} already removed from the superblock");
            }
            Some(inode) => {
```

Reply:
Tried that but it didn't work. I'm afraid it is still locking the `inodes`.

Reply:
I put a similar comment elsewhere, we can avoid the locking with a scoped block.